### PR TITLE
prometheus ssl changes

### PR DESCRIPTION
**Pull request by ananthugnair1**
## Description

This update introduces TLS support for secure connections to Prometheus within the application. The changes include enabling TLS, configuring trust store paths, and handling SSL-related exceptions.
Key Changes:
TLS Configuration:

Added configuration properties to enable TLS:
```
prometheus.tls.enabled=<true/false>
prometheus.tls.truststore-path=<path to truststore>
prometheus.tls.truststore-password=<truststore password>
verify-host-name=<true/false>
```

Conditional logic is implemented to check if TLS is enabled and whether hostname verification is required.

Introduced a custom HostnameVerifier that can be optionally used to bypass hostname verification if verify-host-name is set to false.
## Motivation and Context

Currently, the Prometheus connector in Presto lacks secure communication, making it vulnerable when dealing with sensitive metrics and data. These changes are motivated by the need to enhance security by enabling TLS, ensuring encrypted data exchange, and reducing the risk of unauthorized access. Additionally, improved SSL exception handling with clear error messages increases the reliability of the application, simplifying the diagnosis and resolution of security issues.

Reference issue: https://github.com/prestodb/presto/issues/23555

## Impact

SSL will be enabled and connections can be secured.
## Test Plan

Tested Locally end to end
## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==
Prometheus Docs
* Add Connection security documentation for the Prometheus Connector.

General Changes
* ... :pr:`12345`
* ... :pr:`12345`

Hive Connector Changes
* ... :pr:`12345`
* ... :pr:`12345`
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by steveburnett** [comment]
Suggest adding documentation for the new configuration properties to https://prestodb.io/docs/current/connector/prometheus.html#configuration-properties. 

Suggest adding a release note entry about the new configuration properties, linking to the new documentation.

**Comment by ananthugnair1** [comment]
Hi @steveburnett, 
Updated documentation, please review.
cc: @tdcmeehan

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 1]
```diff
@@ -174,4 +239,33 @@ private Optional<String> getBearerAuthInfoFromFile()
...
+    public SSLContext getSSLContext()
+            throws NoSuchAlgorithmException, KeyStoreException, KeyManagementException
+    {
+        SSLContext sslContext = SSLContext.getInstance("TLS");
+        sslContext.init(null, getTrustManagerFactory().getTrustManagers(), new java.security.SecureRandom());
+        return sslContext;
+    }
+
+    public TrustManagerFactory getTrustManagerFactory()
+            throws KeyStoreException, NoSuchAlgorithmException
```
Comment (elharo) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:252]: can this throw NoSuchAlgorithmException? It's caught below.

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusConnectorConfig.java [hunk 2]
```diff
@@ -115,4 +119,49 @@ public void checkConfig()
...
+        return truststorePassword;
+    }
+
+    @Config("prometheus.tls.truststore-password")
+    public PrometheusConnectorConfig setTruststorePassword(String password)
+    {
+        this.truststorePassword = password;
+        return this;
+    }
+    public boolean getVerifyHostName()
```
Comment (elharo) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusConnectorConfig.java:157]: nit: add blank line

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusConnectorConfig.java [hunk 3]
```diff
@@ -115,4 +119,49 @@ public void checkConfig()
...
+    {
+        return verifyHostName;
+    }
+    @Config("verify-host-name")
```
Comment (elharo) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusConnectorConfig.java:161]: blank line

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java (hunk 1 above)
Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:252]: TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm()); at line 267 throws NoSuchAlgorithmException

Comment (elharo) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:252]: perhaps move this into the try block then?

Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:252]: Hi @elharo ,
Moved to try block, please review

Comment (elharo) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:252]: remove from throws clause then.
also consider if you want to convert KeyStoreException to PrestoException here.

Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java]: removed NoSuchAlgorithmException from throws clause.
KeyStoreException is converting to PrestoException on line 221
        catch (KeyStoreException e) {
            throw new PrestoException(KEYSTORE_OPERATION_ERROR, "Keystore operation error", e);
        }

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pull branch, local doc build, looks good. Thanks!

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java (hunk 1 above)
Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java]: Hi @elharo could you review this?

Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java]: hi @elharo could you please continue this review?

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 4]
```diff
@@ -148,17 +169,61 @@ public byte[] fetchUri(URI uri)
...
+                if (response.isSuccessful() && response.body() != null) {
+                    return response.body().bytes();
+                }
             }
         }
+        catch (SSLHandshakeException e) {
+            throw new PrestoException(SSL_HANDSHAKE_ERROR, "An SSL handshake error occurred while establishing a secure connection. Try the following measures to resolve the error:\n\n" + "- Upload a valid SSL certificate for authentication\n- Verify the expiration status of the uploaded certificate.\n- If you are connecting with SSL, enable SSL on both ends of the connection.\n");
+        }
+        catch (SSLPeerUnverifiedException e) {
+            throw new PrestoException(PEER_VERIFY_ERROR, "Peer verification failed.Below measures may resolve the issue \n" +
```
Comment (elharo) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:211]: missing space after period
Below --> These
may --> might (I think, never sure about that one)

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 5]
```diff
@@ -148,17 +169,61 @@ public byte[] fetchUri(URI uri)
...
+                    "- Add correct Hostname in the SSL certificate's SAN list \n" +
+                    "- The certificate chain might be incomplete. Please check your SSL certificate\n");
```
Comment (elharo) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:213]: No Please

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 6]
```diff
@@ -174,4 +239,33 @@ private Optional<String> getBearerAuthInfoFromFile()
             }
         });
     }
+
+    public SSLContext getSSLContext()
```
Comment (elharo) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:243]: does this method need to be public?

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 7]
```diff
@@ -148,17 +169,61 @@ public byte[] fetchUri(URI uri)
...
         getBearerAuthInfoFromFile().map(bearerToken -> requestBuilder.header("Authorization", "Bearer " + bearerToken));
-
         Response response;
         try {
-            response = httpClient.newCall(requestBuilder.build()).execute();
-            if (response.isSuccessful() && response.body() != null) {
-                return response.body().bytes();
+            if (config.isTlsEnabled()) {
+                OkHttpClient httpClient;
+                HostnameVerifier hostnameVerifier = new HostnameVerifier() {
```
Comment (tdcmeehan) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:176]: Can't this be a lambda?

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 8]
```diff
@@ -57,12 +76,14 @@ public class PrometheusClient
...
     private final Supplier<Map<String, Object>> tableSupplier;
     private final Type varcharMapType;
+    private PrometheusConnectorConfig config;
 
     private static final Logger log = Logger.get(PrometheusClient.class);
 
     @Inject
     public PrometheusClient(PrometheusConnectorConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
     {
+        this.config = config;
```
Comment (tdcmeehan) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:86]:
```suggestion
        this.config = requireNonNull(config, "config is null");
```

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 9]
```diff
@@ -148,17 +169,61 @@ public byte[] fetchUri(URI uri)
...
+            }
+            else {
+                response = httpClient.newCall(requestBuilder.build()).execute();
+                if (response.isSuccessful() && response.body() != null) {
+                    return response.body().bytes();
+                }
             }
         }
+        catch (SSLHandshakeException e) {
+            throw new PrestoException(SSL_HANDSHAKE_ERROR, "An SSL handshake error occurred while establishing a secure connection. Try the following measures to resolve the error:\n\n" + "- Upload a valid SSL certificate for authentication\n- Verify the expiration status of the uploaded certificate.\n- If you are connecting with SSL, enable SSL on both ends of the connection.\n");
```
Comment (tdcmeehan) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:208]: Embed the exception

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java [hunk 10]
```diff
@@ -148,17 +169,61 @@ public byte[] fetchUri(URI uri)
...
+        }
+        catch (SSLPeerUnverifiedException e) {
```
Comment (tdcmeehan) [inline presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:210]: Embed the exception

File: presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java [hunk 11]
```diff
@@ -118,6 +119,12 @@ public enum StandardErrorCode
     UNSUPPORTED_ANALYZER_TYPE(0x0001_0029, INTERNAL_ERROR),
     SLICE_TOO_LARGE(0x0001_002A, INTERNAL_ERROR),
     HBO_FAILURE(0x0001_002B, INTERNAL_ERROR),
+    CRYPT_ALGORITHM_ERROR(0x0001_002C, EXTERNAL),
+    CERT_PARSE_ERROR(0x0001_002D, EXTERNAL),
+    PEER_VERIFY_ERROR(0x0001_002E, EXTERNAL),
+    SSL_HANDSHAKE_ERROR(0x0001_002F, EXTERNAL),
+    KEY_MANAGEMENT_ERROR(0x0001_0030, EXTERNAL),
+    KEYSTORE_OPERATION_ERROR(0x0001_0031, EXTERNAL),
```
Comment (tdcmeehan) [inline presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java:127]: I think we can group these into a single error category related to SSL connection managment.  Please check if one exists that we can use already.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Any way to add an end to end test?

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java (hunk 10 above)
Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:210]: done

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java (hunk 9 above)
Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:208]: done

File: presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java (hunk 7 above)
Comment (ananthugnair1) [reply presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java:176]: changed to lambda

**Comment by tdcmeehan** [comment]
Please squash commits

**Comment by ananthugnair1** [comment]
> Please squash commits

Squashed commits

**Comment by tdcmeehan** [comment]
Let's word commit like so: `Add TLS support for Prometheus connector`

**Comment by ananthugnair1** [comment]
> Let's word commit like so: `Add TLS support for Prometheus connector`

Changed commit message.

**Review by tdcmeehan — APPROVED** [review]

**Pull request merged**